File: kmp/cpp/kmp.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

std::vector<int> 
LPSTable(const std::string& pattern) 
{
    std::vector<int> lps(pattern.size(), 0);
    long long int len = 0;
    long long int i = 1;

    while (i < pattern.size()) 
    {
        if (pattern[i] == pattern[len]) 
        {
            lps[i] = ++len;
            ++i;
        }
        else
        {
            if (len != 0)
            {
                len = lps[len - 1];
            }
            else
            {
                lps[i] = 0;
                ++i;
            }
        }
    }

    return lps;
}

void 
KMPSearch(const std::string& pattern, const std::string& filename)
{
    std::ifstream input_file(filename);
    if (!input_file)
    {
        std::cout << "Failed to open the input file." << std::endl;
        return;
    }

    std::vector<int> lps = LPSTable(pattern);
    std::string line;
    int line_index = 1;

    while (std::getline(input_file, line))
    {
        int i = 0;
        long long int j = 0;

        while (i < line.size())
        {
            if (pattern[j] == line[i])
            {
                ++i;
                ++j;
            }

            if (j == pattern.size())
            {
                std::cout << "Pattern found at line " << line_index 
                          << ", index " << (i - j) << std::endl;
                j = lps[j - 1];
            }
            else if (i < line.size() && pattern[j] != line[i])
            {
                if (j != 0)
                {
                    j = lps[j - 1];
                }
                else
                {
                    ++i;
                }
            }
        }

        ++line_index;
    }

    input_file.close();
}

int main()
{
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);
    
    std::string pattern;
    std::cout << "Enter the pattern to search: ";
    std::getline(std::cin, pattern);

    std::string filename = "kmp_dna_input.txt";
    
    KMPSearch(pattern, filename);

    return 0;
}

File: kmp/python/kmp.py
def LPSTable(pattern):
    lps = [0] * len(pattern)
    length = 0
    i = 1

    while i < len(pattern):
        if pattern[i] == pattern[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            if length != 0:
                length = lps[length - 1]
            else:
                lps[i] = 0
                i += 1

    return lps

def KMPSearch(pattern, filename):
    try:
        with open(filename, 'r') as input_file:
            lps = LPSTable(pattern)
            line_index = 1

            for line in input_file:
                i, j = 0, 0

                while i < len(line):
                    if pattern[j] == line[i]:
                        i += 1
                        j += 1

                    if j == len(pattern):
                        print(f"Pattern found at line {line_index}, index {i - j}")
                        j = lps[j - 1]
                    elif i < len(line) and pattern[j] != line[i]:
                        if j != 0:
                            j = lps[j - 1]
                        else:
                            i += 1

                line_index += 1
    except FileNotFoundError:
        print("Failed to open the input file.")

if __name__ == "__main__":
    pattern = input("Enter the pattern to search: ")
    filename = "kmp_lorem_ipsum_input.txt"
    KMPSearch(pattern, filename)